### PR TITLE
Introduce teamraum document final state

### PR DESCRIPTION
**Pull request by Acrop146**
Before: When a document was finalized in Gever and then copied to Teamroom, it lost its “finalized” status could be edited in the Teamroom. During upload back to Gever, a new version was created in the actually finalized document.

To fix this problem - a final state was created for Workspace documents (transitions not accessible to normal users). Additionally, overwriting a finalized Gever Document from Teamraum now throws an error.

For [TI-21]

## Checklist

- [x] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)

- Upgrade-Steps:

  - [x] Make it deferrable if possible
  - [x] Execute as much as possible conditionally

- New translations
  - [x] All msg-strings are unicode
- Change in schema definition:
  - [x] If `missing_value` is specified, then `default` has to be set to the same value

[TI-21]: https://4teamwork.atlassian.net/browse/TI-21?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: opengever/api/linked_workspaces.py [hunk 1]
```diff
@@ -455,7 +455,7 @@ def reply(self):
         except CopyFromWorkspaceForbidden:
             raise BadRequest(
                 _("Document can't be copied from workspace because it's "
-                  "currently checked out"))
+                  "currently checked out or Gever Document is finalized"))
```
[the remark this answers is not in the document]
Comment (elioschmutz) [reply opengever/api/linked_workspaces.py:458]: generally, this the message-id which should not be the english translation... but it was already wrong.

File: opengever/workspaceclient/linked_workspaces.py [hunk 2]
```diff
@@ -367,6 +367,12 @@ def copy_document_from_workspace(
         # for example because the GEVER document was trashed, always create
         # a copy instead of attempting to create a version.
         gever_doc = self._get_corresponding_gever_doc(document_repr)
+
+        if gever_doc.is_final_document():
+            raise CopyFromWorkspaceForbidden(
```
Comment (elioschmutz) [inline opengever/workspaceclient/linked_workspaces.py:372]: I'd use a new error for the different errors or use some kind of an exception attribute to differentiate it. After that, you can write more specific error messages for the user and there is no need of combining the messages.

File: changes/TI-21.feature [hunk 3]
```diff
@@ -0,0 +1 @@
+Add new workspace document state: opengever_workspace_document--STATUS--final" title="finalize. [ran]
```
Comment (elioschmutz) [inline changes/TI-21.feature:1]: ✂️  ` title="finalize.`

**Review by elioschmutz — CHANGES_REQUESTED** [review]
There seems to be a little mess with the commit order and commit content. Could you please update the PR and clean up the commits?

And please, use a meaningful PR instead of your nickname and jira-number. So i.e. `Introduce teamraum document final state`.

File: opengever/workspaceclient/linked_workspaces.py [hunk 4]
```diff
@@ -288,6 +288,7 @@ def copy_document_to_workspace(self, document, workspace_uid, lock=False, folder
             return self.client.post(target_url, json=document_repr)
 
         document_metadata['gever_url'] = self.client.get_gever_url(Oguid.for_object(document).id)
+        document_metadata['final'] = document.is_final_document()
```
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/workspaceclient/linked_workspaces.py:291]: Yes, review state isnt present in "document_metadata" when I list it using "pp document_metadata

File: opengever/base/locales/de/LC_MESSAGES/plone.po [hunk 5]
```diff
@@ -177,9 +177,6 @@ msgstr "nicht definiert"
 msgid "document-state-final"
 msgstr "Final"
 
-msgid "opengever_workspace_document--STATUS--final"
-msgstr "Final"
-
```
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/base/locales/de/LC_MESSAGES/plone.po]: Added to first commit

File: opengever/api/workspace.py [hunk 6]
```diff
@@ -80,3 +81,21 @@ def before_deserialization(self, obj):
...
+            chain = wftool.getChainFor(self.obj)
+            workflow_id = chain[0]
+            wftool.setStatusOf(workflow_id, self.obj, {
+                'review_state': self.obj.final_state_workspace,
+                'action': '',
+                'actor': ''})
+            workflow = wftool.getWorkflowById(workflow_id)
+            workflow.updateRoleMappingsFor(self.obj)
+            self.obj.reindexObject(idxs=['review_state'])
+            self.obj.reindexObjectSecurity()
```
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/api/workspace.py:101]: Implemented in "as_final_document" function.

File: opengever/api/linked_workspaces.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/api/linked_workspaces.py:458]: Understood!

File: opengever/core/lawgiver.zcml [hunk 7]
```diff
@@ -104,6 +104,19 @@
                    opengever.task: Edit task,
                    "
       />
+  <lawgiver:map_permissions
```
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/core/lawgiver.zcml:107]: Done, Lawgiver and all the resulting translations (when lawgiver is implemented) have been moved to first commit.

File: opengever/api/workspace.py [hunk 8]
```diff
@@ -84,6 +84,7 @@ def before_serialization(self, obj):
 
     def add_object_to_context(self):
 
+        super(UploadDocumentCopy, self).add_object_to_context()
```
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/api/workspace.py:87]: Done, moved to first commit

File: opengever/api/tests/test_linked_workspaces.py [hunk 9]
```diff
@@ -1752,9 +1752,9 @@ def test_copying_document_from_workspace_is_prevented_if_checked_out(self, brows
                 {u'type': u'BadRequest',
                  u'additional_metadata': {},
                  u'message': u"Document can't be copied from workspace "
-                             u"because it's currently checked out",
+                             u"because it's currently checked out or Gever Document is finalized",
```
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/api/tests/test_linked_workspaces.py:1755]: Done, squashed the commit into the first one.

File: opengever/workspaceclient/linked_workspaces.py (hunk 2 above)
Comment (Acrop146) [reply opengever/workspaceclient/linked_workspaces.py]: Done, implemented as new exception: GeverDocumentCantBeChanged

File: opengever/core/profiles/default/workflows/opengever_workspace_document/specification.txt [hunk 10]
```diff
@@ -35,3 +36,12 @@ General:
 Initial Status: Active
 Status Active:
   A workspace guest can view.
+  A workspace member can edit workspace document.
+  A workspace admin can finalize.
```
Comment (elioschmutz) [inline opengever/core/profiles/default/workflows/opengever_workspace_document/specification.txt:40]: is there a reason why we do restrict finalizing to admins only?

**Review by elioschmutz — APPROVED** [review]
👍 thanks! 

please see my comments before you merge

File: opengever/core/profiles/default/workflows/opengever_workspace_document/specification.txt (hunk 10 above)
Comment (Acrop146) [reply opengever/core/profiles/default/workflows/opengever_workspace_document/specification.txt:40]: Yes, the story mentioned that "normal users" shouldn't be able to transition the state of document

File: opengever/api/workspace.py [hunk 11]
```diff
@@ -80,3 +80,11 @@ def before_deserialization(self, obj):
 
     def before_serialization(self, obj):
         ILinkedDocuments(obj).link_gever_document(self.gever_document_uid)
+
+    def add_object_to_context(self):
+
```
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/api/workspace.py:85]: Done

File: opengever/api/workspace.py [hunk 12]
```diff
@@ -80,3 +80,11 @@ def before_deserialization(self, obj):
...
+        super(UploadDocumentCopy, self).add_object_to_context()
+        data = json.loads(self.request.form['document_metadata'])
+
+        if self.obj.portal_type == 'opengever.document.document' and data.get("final", None):
```
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/api/workspace.py:89]: Implemented

File: opengever/document/document.py [hunk 13]
```diff
@@ -397,6 +398,23 @@ def as_shadow_document(self):
...
 
+    def as_final_document(self):
+        """Force a workspace document into the final state.
+
+        """
+        wftool = api.portal.get_tool('portal_workflow')
+        chain = wftool.getChainFor(self)
+        workflow_id = chain[0]
+        wftool.setStatusOf(workflow_id, self, {
+            'review_state': self.final_state_workspace,
```
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/document/document.py:409]: Implemented as:
`        if workflow_id == self.workspace_workflow_id:
            wftool.setStatusOf(workflow_id, self, {
            'review_state': self.final_state_workspace,
            'action': '',
            'actor': ''})
        else:
            wftool.setStatusOf(workflow_id, self, {
                'review_state': self.final_state,
                'action': '',
                'actor': ''})`

File: opengever/document/document.py [hunk 14]
```diff
@@ -502,7 +520,7 @@ def is_shadow_document(self):
         return api.content.get_state(self) == self.shadow_state
 
     def is_final_document(self):
-        return api.content.get_state(self) == self.final_state
+        return api.content.get_state(self) == self.final_state or api.content.get_state(self) == self.final_state_workspace
```
[the remark this answers is not in the document]
Comment (Acrop146) [reply opengever/document/document.py]: Done

**Pull request merged**